RFC: decommissioning a node

neeral opened the pull request:
@tschottdorf

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=16352) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=16352) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: Neeral Dodhia<br/>:x: neeral<br/><hr/>**Neeral Dodhia** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cockroachdb/cockroach?pullRequest=16352) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16352)
<!-- Reviewable:end -->

neeral commented:
RFC is entering final comment period

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/decommission.md, line 30 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Kly3_chDF-tDagNDGNg:-Kly3_chDF-tDagNDGNh:b-9cerel) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L30)):*
> ```Markdown
> 1. Change allocator to treat decommissioned nodes like dead nodes with the exception that we will down-replicate even if there is no target for up-replication.
>    1. All replicas for each replica set are up-replicated.
>    1. All replicas deleted from the node.
> ```


1. Node or coordinating process tracks decommissioning status. When completed (i.e. node has shed all of its data and is ready to be terminated), perform cosmetic cleanup to avoid showing this node as dead permanently (see further below).

---

*[docs/RFCS/decommission.md, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Kly4-gnLFGNsDK68psi:-Kly4-gnLFGNsDK68psj:b-ilxvu5) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L41)):*
> ```Markdown
> - Leaseholder of ranges with replicas on target node trigger up-replication.
> 
> ### Difference in process when target is alive vs dead
> ```

Can you consistently wrap the document at, say, 80 or 100 chars?

---

*[docs/RFCS/decommission.md, line 42 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Kly405_6pNPbZhxlJvR:-Kly405_6pNPbZhxlJvS:b-rkx2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L42)):*
> ```Markdown
> 
> ### Difference in process when target is alive vs dead
> If the node being removed is dead, and so, unreachable, its leases and data would already have been transferred to other nodes. The only thing to do is preventing it from rejoining the cluster if it were to become available. This would require blacklisting that node and would be fairly complex: this will not be attempted.
> ```

I think what we can propose here is that you at least get to run the cosmetic command that removes the node from the UI (at least until the node comes back - which it never should) and from relevant views where it would otherwise remain, with a decided emphasis on being cosmetic and without "gotchas" (we don't want an accidental cosmetic removal of a node that is actually live or coming back to have real consequences).

---

*[docs/RFCS/decommission.md, line 45 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Kly2juW3hroX1aE23gT:-Kly2juW3hroX1aE23gU:b-w7piw3) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L45)):*
> ```Markdown
> 
> ## Temporary removal
> No changes required. The existing CockroachDb process, described below, is sufficient.
> ```

s/Db/DB/ throughout the doc.

---

*[docs/RFCS/decommission.md, line 54 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Kly3AXKlLOYbTK-8orG:-Kly3AXKlLOYbTK-8orH:bnaia1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L54)):*
> ```Markdown
> 
> # Unresolved questions
> - Interface for initiating removal of a node. One suggestion is adding `node decommission <node-id>` to the CLI. The operation should be asynchronous (vs blocking). If it is blocking and it takes a long time to complete, connection might timeout if no response was sent to the client.
> ```

I think you can just move this into the main proposal and make it more assertive.

---

*[docs/RFCS/decommission.md, line 55 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Kly3DVdf91RiBOGxwXw:-Kly3DVdf91RiBOGxwXx:bslzko) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L55)):*
> ```Markdown
> # Unresolved questions
> - Interface for initiating removal of a node. One suggestion is adding `node decommission <node-id>` to the CLI. The operation should be asynchronous (vs blocking). If it is blocking and it takes a long time to complete, connection might timeout if no response was sent to the client.
> - Removing multiple nodes concurrently
> ```

This is also straightforward. The cli command can take multiple endpoints.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-:-Kly4XJH5H2uk5GGZ2Lj:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/decommission.md, line 34 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-KlyDK0zDyR7d9TL_MuP:-KlyDK1-lAno390tAauU:b-ku04g1) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L34)):*
> ```Markdown
> 
> Lower-level changes:
> - Add `Decommissioned`, a boolean field, to `StoreIdent`. The `StoreIdent` is persisted to disk.
> ```

If the node crashes at this point, we won't be able to restart it in order to continue to drain data from it. This can be problematic if there are replicas on the node which are required for quorum on their associated ranges. Either the `Decommissioned` state should be persisted after the node is drained of data, or there should be a two-step dance here where a node is marked as `Decommissioning`, then we drain data, then we mark as `Decommissioned`.

---

*[docs/RFCS/decommission.md, line 38 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-KlyEUFFP7oJeuiekIvF:-KlyEUFFP7oJeuiekIvG:b-ll63km) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L38)):*
> ```Markdown
> - Through the gossip network, all nodes are informed that node is being removed via updated `StoreDescriptor` message.
> - Prevent any new ranges being allocated to target node.
> - Transfer leases away from target node.
> ```

Note that transferring leases away from a draining node is already present. See `Store.SetDraining`. Also, we mark a node for draining via `Liveness.Draining`. Perhaps we should add `Liveness.Decommissioning` as well. I mention this because it might fit in better with the `StorePool` which already pays attention to `Liveness.Draining`. 

---

*[docs/RFCS/decommission.md, line 54 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Kly3AXKlLOYbTK-8orG:-KlyEtYOseZYwFS42qP7:brfaekv) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L54)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I think you can just move this into the main proposal and make it more assertive.
</blockquote></details>

Bikeshed: perhaps name the command`node rm`. This fits in better with other CLI commands such as `user rm` and `zone rm`. 

Probably want to prompt the user for confirmation (with a command line flag to skip confirmation).

We'll want to provide some mechanism to determine that the decommissioning has finished (e.g. via `node ls`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-:-KlyG-JFPsx3PALVyx8L:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/decommission.md, line 28 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km-KaobJDVxpyjcwxn_:-Km-KaobJDVxpyjcwxna:b-dd0ug6) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L28)):*
> ```Markdown
> 1. Marks that it is being decommissioned.
> 1. Node does not renew its leases causing all them to be transferred.
> 1. Change allocator to treat decommissioned nodes like dead nodes with the exception that we will down-replicate even if there is no target for up-replication.
> ```

I don't think it's a good idea to down-replicate even if there is no up-replication target. Better to stall the decommissioning process until new nodes are available (and to bring this to the operator's attention so that further decommisionings are not attempted)

---

*[docs/RFCS/decommission.md, line 34 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-KlyDK0zDyR7d9TL_MuP:-Km-LDD-JVAcC74csxtL:b-shhuhs) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L34)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

If the node crashes at this point, we won't be able to restart it in order to continue to drain data from it. This can be problematic if there are replicas on the node which are required for quorum on their associated ranges. Either the `Decommissioned` state should be persisted after the node is drained of data, or there should be a two-step dance here where a node is marked as `Decommissioning`, then we drain data, then we mark as `Decommissioned`.
</blockquote></details>

I'm not sure we need an additional state. The decommissioning state can be a one-way boolean flag. Once it's been set, the node will not accept any new replicas and will attempt to shed the ones it has. As long as it has replicas, it may restart in order to participate in quorums for its own down-replication. Once it is decommissioning *and* has no more replicas, it may refuse to restart. (although I'm not sure about this as a design - i think it may be better to let the process start but in a defunct state where it cannot receive any more replicas)

---

*[docs/RFCS/decommission.md, line 35 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km-LsavPRjnJfonRLSl:-Km-LsavPRjnJfonRLSm:btkw2bs) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L35)):*
> ```Markdown
> Lower-level changes:
> - Add `Decommissioned`, a boolean field, to `StoreIdent`. The `StoreIdent` is persisted to disk.
> - When a node bootstraps, it checks the `StoreIdent`s for each of its stores. If any of the StoreIdents have `Decommissioned` set to True, then kill the process with an appropriate error message like "Attempted to start a decommissioned node. Please wipe stores and retry".
> ```

Also propagate the decommissioned flag to any stores that may not have it (similar to the way that we propagate the cluster ID to new stores at startup)

---

*[docs/RFCS/decommission.md, line 42 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km-MB1kj2am05Tywl7T:-Km-MB1kj2am05Tywl7U:b8icgw3) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L42)):*
> ```Markdown
> 
> ### Difference in process when target is alive vs dead
> If the node being removed is dead, and so, unreachable, its leases and data would already have been transferred to other nodes. The only thing to do is preventing it from rejoining the cluster if it were to become available. This would require blacklisting that node and would be fairly complex: this will not be attempted.
> ```

Its leases would be transferred within a few seconds; its data may take longer depending on the availability of up-replication targets. The tool that manages the decommissioning process needs to be able to watch the range descriptors to learn when the dead node has been down-replicated from all its ranges. 

---

*[docs/RFCS/decommission.md, line 54 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Kly3AXKlLOYbTK-8orG:-Km-NnkVWySE8DFwA5f4:bij72qn) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L54)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Bikeshed: perhaps name the command`node rm`. This fits in better with other CLI commands such as `user rm` and `zone rm`. 

Probably want to prompt the user for confirmation (with a command line flag to skip confirmation).

We'll want to provide some mechanism to determine that the decommissioning has finished (e.g. via `node ls`).
</blockquote></details>

I think the command should block, but be retryable (so if the command fails, just run `cockroach node rm` again). The command exits when it is safe to wipe/power down the machine. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352)*
<!-- Sent from Reviewable.io -->

neeral commented:
> 100
Test comment

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/decommission.md, line 28 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km-KaobJDVxpyjcwxn_:-Km1khLWJ_A83qEUrSmm:bsvr7i5) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L28)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I don't think it's a good idea to down-replicate even if there is no up-replication target. Better to stall the decommissioning process until new nodes are available (and to bring this to the operator's attention so that further decommisionings are not attempted)
</blockquote></details>

I agree that we  don't need to down-replicate if there is no up-replication target. I think it would be better to reuse the logic for draining. This will keep things simpler. Now that I better understand Rubrik's requirements, removing data from the decommissioned node isn't necessary.

Question for @bdarnell @tschottdorf : is it necessary to stall the process? The draining process isn't stalled. Is there a reason that this would be different?

---

*[docs/RFCS/decommission.md, line 35 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km-LsavPRjnJfonRLSl:-Km1m4WIMXZw6sMhuPJg:bitujcd) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L35)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Also propagate the decommissioned flag to any stores that may not have it (similar to the way that we propagate the cluster ID to new stores at startup)
</blockquote></details>

What's the reason for this? I was assuming that stores can't be added to another node. Is it to make sure all the stores are in the same state?

---

*[docs/RFCS/decommission.md, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Kly4-gnLFGNsDK68psi:-Klz7W7gmQUbL9hxdJJ-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L41)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Can you consistently wrap the document at, say, 80 or 100 chars?
</blockquote></details>

Done.

---

*[docs/RFCS/decommission.md, line 42 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km-MB1kj2am05Tywl7T:-Km1mXasvgYoaU6QhypT:bik23tj) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L42)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Its leases would be transferred within a few seconds; its data may take longer depending on the availability of up-replication targets. The tool that manages the decommissioning process needs to be able to watch the range descriptors to learn when the dead node has been down-replicated from all its ranges. 
</blockquote></details>

Toby and I were thinking that the node tracks its decommission process. When all its replicas have been up-replicated to another node, it will make the cosmetic change to remove itself from the node table in admin UI. This event can be observed by the tool. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-:-KlysvkFFAhZHwmLJ1gp:b-ad1azr)*
<!-- Sent from Reviewable.io -->

neeral commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/decommission.md, line 28 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km-KaobJDVxpyjcwxn_:-Km2v61_tNyb3rkhm8eu:b-bbht3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L28)):*
<details><summary><i>Previously, neeral (Neeral Dodhia) wrote…</i></summary><blockquote>

I agree that we  don't need to down-replicate if there is no up-replication target. I think it would be better to reuse the logic for draining. This will keep things simpler. Now that I better understand Rubrik's requirements, removing data from the decommissioned node isn't necessary.

Question for @bdarnell @tschottdorf : is it necessary to stall the process? The draining process isn't stalled. Is there a reason that this would be different?
</blockquote></details>

Ignore my comment above. The process will stall to guarantee a minimum number of replicas for each replica set.

---

*[docs/RFCS/decommission.md, line 34 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-KlyDK0zDyR7d9TL_MuP:-Km2uXSe_Dj1BWIduF8M:b-tcxkov) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L34)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'm not sure we need an additional state. The decommissioning state can be a one-way boolean flag. Once it's been set, the node will not accept any new replicas and will attempt to shed the ones it has. As long as it has replicas, it may restart in order to participate in quorums for its own down-replication. Once it is decommissioning *and* has no more replicas, it may refuse to restart. (although I'm not sure about this as a design - i think it may be better to let the process start but in a defunct state where it cannot receive any more replicas)
</blockquote></details>

We'll go with Ben's approach of "decommissioning and has no more replicas" to determine that decommissioning has finished

---

*[docs/RFCS/decommission.md, line 35 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km-LsavPRjnJfonRLSl:-Km2uhU83odfP2ag9CpN:bysotnq) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L35)):*
<details><summary><i>Previously, neeral (Neeral Dodhia) wrote…</i></summary><blockquote>

What's the reason for this? I was assuming that stores can't be added to another node. Is it to make sure all the stores are in the same state?
</blockquote></details>

Not necessary anymore - we won't be writing to StoreIdents

---

*[docs/RFCS/decommission.md, line 42 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Kly405_6pNPbZhxlJvR:-Km2unOYQW_NFa8eZ8y-:bma6in6) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L42)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I think what we can propose here is that you at least get to run the cosmetic command that removes the node from the UI (at least until the node comes back - which it never should) and from relevant views where it would otherwise remain, with a decided emphasis on being cosmetic and without "gotchas" (we don't want an accidental cosmetic removal of a node that is actually live or coming back to have real consequences).
</blockquote></details>

@tschottdorf please read the Unanswered Questions part of the latest revision to the PR. At least according to my understanding, there is no need for this cosmetic change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

It looks like you need to squash everything into a single commit with the correct author email address to pass the CLA check. 

---

Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/decommission.md, line 42 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km-MB1kj2am05Tywl7T:-Km35-7Z1cxDL5djtlql:b-v74ytx) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L42)):*
<details><summary><i>Previously, neeral (Neeral Dodhia) wrote…</i></summary><blockquote>

Toby and I were thinking that the node tracks its decommission process. When all its replicas have been up-replicated to another node, it will make the cosmetic change to remove itself from the node table in admin UI. This event can be observed by the tool. 
</blockquote></details>

The node cannot always remove itself from the admin UI, because it may be dead. I think it's best if other nodes are always responsible for removing decommissioned nodes from the admin UI. 

---

*[docs/RFCS/decommission.md, line 78 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km34RQ5fmtYITHCvGdo:-Km34RQ5fmtYITHCvGdp:b-cg4vjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee9d6a23e11b7d157725bd45bf860116b393869/docs/RFCS/decommission.md#L78)):*
> ```Markdown
> need to prevent a decommissioned node from restarting.
> 
> When a node restarts, there is a small period when it can accept new replicas
> ```

It's also worth noting that other nodes will not knowingly send new replicas to a decommissioned node, so this would require both nodes to be unaware of the decommissioning status (which should be rare and short-lived). 

---

*[docs/RFCS/decommission.md, line 147 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km35RNZ809C-MKFDd4A:-Km35RNZ809C-MKFDd4B:b-dy701c) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee9d6a23e11b7d157725bd45bf860116b393869/docs/RFCS/decommission.md#L147)):*
> ```Markdown
> cluster, two users can simultaneously request to decommission two different
> nodes each. The resulting state leaves two nodes in decommissioning state and
> only one live node. A quorum cannot be reached and so progress halts. In this
> ```

Decommissioning nodes can still participate in quorums. The replication changes cannot make progress because there are not a sufficient number of non-decommissioned nodes, but that's separate from an inability to reach a quorum.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-:-Km35qyP08nWBVl8EWMd:b-s4l5p8)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> just some minor points and what Ben said.

---

Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/decommission.md, line 24 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km3F7Khp7Svax_12-dR:-Km3F7Khp7Svax_12-dS:b-tyjbox) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee9d6a23e11b7d157725bd45bf860116b393869/docs/RFCS/decommission.md#L24)):*
> ```Markdown
> re-replicating previously deleted entries. CockroachDB does not suffer from this
> specific issue.
> ```

Motivate the support for decommissioning multiple nodes at once: Doing it serially inflicts a lot more data movement as data moves off one soon-to-be-dead-node to another and on and on.

---

*[docs/RFCS/decommission.md, line 53 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km3EHSarOvX5oiYJAX7:-Km3EHSarOvX5oiYJAX8:b-p3pl4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee9d6a23e11b7d157725bd45bf860116b393869/docs/RFCS/decommission.md#L53)):*
> ```Markdown
>    1. is not accepting new replicas (TODO check whether this is the case)
> 1. Leaseholders, necessarily on non-target nodes, for ranges where a target node
>    is a member of the replica set:
> ```

Mention that the replicate queue will be taught to treat decommissioning nodes much like dead replicas. That will do the right thing: we don't downreplicate if that's putting us in more danger than we already are, and we won't upreplicate to more dangerous states (Garvit is working on that).

---

*[docs/RFCS/decommission.md, line 65 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km3EpQ39ctaWdjYrWZX:-Km3EpQ39ctaWdjYrWZY:b-faw1nb) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee9d6a23e11b7d157725bd45bf860116b393869/docs/RFCS/decommission.md#L65)):*
> ```Markdown
>    for each target node. This explicitly shuts down the node. Otherwise, the
>    node remains up but is unable to participate in usual operations such as
>    having replicas. The `--decommission` flag causes it to wait until the
> ```

Add that `quit --decommission` also (idempotently) initiates decommission, the difference being that it blocks until the node shuts down. This means users that don't want an asynchronous external polling while the data is draining can call just that (at least if all they want to decommission is a single node).

---

*[docs/RFCS/decommission.md, line 91 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km3FmAzkMhzn_3_F5_f:-Km3FmAzkMhzn_3_F5_g:boqzbrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee9d6a23e11b7d157725bd45bf860116b393869/docs/RFCS/decommission.md#L91)):*
> ```Markdown
> participate like normal.
> Node restart is required because we cannot determine whether a node is in
> `Draining` because if was previously in `Decommissioning` or for another reason.
> ```

I think you *could* do better here. If the node is live, you could directly tell it to undrain. But I think it's OK to punt on this for now. Perhaps make a note.

---

*[docs/RFCS/decommission.md, line 104 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km3GQOTM0aOqGoCPneD:-Km3GQOTM0aOqGoCPneE:beb787a) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee9d6a23e11b7d157725bd45bf860116b393869/docs/RFCS/decommission.md#L104)):*
> ```Markdown
> because it is dead. Instead, its leases will expire and will be taken by other
> nodes. After `COCKROACH_TIME_UNTIL_STORE_DEAD` elapses, the replicas of its
> ranges will actively be rebalanced. Wait for the replica count to reach 0 as per
> ```

s/per/for/?

---

*[docs/RFCS/decommission.md, line 133 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km3GdDuPVg52b1UrRlh:-Km3GdDuPVg52b1UrRli:b-7zmkas) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee9d6a23e11b7d157725bd45bf860116b393869/docs/RFCS/decommission.md#L133)):*
> ```Markdown
>   all nodes, their statuses, replica count, decommissioning flag and draining
>   flag. If the number of nodes remaining after decommissioning is fewer than
>   required for a quorum, the command will return early without updating state.
> ```

This isn't so trivial due to ZoneConfig, etc, but we should obviously warn whenever we get any information. One such criterion is `number_of_remaining_nodes < max(zone.number_of_replicas_desired)` but additionally you may have to satisfy constraints which makes that check too optimistic. I'd be comfortable moving this to a "nice to have" section because it may actually backfire when the command looks "safe" but isn't. (The case we explicitly discussed in our whiteboard session was whether it should be allowed to try to decommission all nodes, which effectively makes the cluster unavailable for clients, and the consensus was that if that's what you do, that's what you get).

---

*[docs/RFCS/decommission.md, line 147 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km35RNZ809C-MKFDd4A:-Km3HYHbywgPinQaXLr9:bebd1px) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee9d6a23e11b7d157725bd45bf860116b393869/docs/RFCS/decommission.md#L147)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Decommissioning nodes can still participate in quorums. The replication changes cannot make progress because there are not a sufficient number of non-decommissioned nodes, but that's separate from an inability to reach a quorum.
</blockquote></details>

Yeah, that's not really a drawback of the algorithm here unless you're trying to make all invocations of `decommission` either succeed in bounded time or fail. The cluster will simply refuse to decommission itself to death, and decommissioning will hang. Nice to detect, but with an effort that I think is disproportionate related to the win it gives in the short term.

---

*[docs/RFCS/decommission.md, line 151 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km3Hu7rZd29yzfSwcTq:-Km3Hu7sGf7apQS2Egh_:b-1lb70s) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee9d6a23e11b7d157725bd45bf860116b393869/docs/RFCS/decommission.md#L151)):*
> ```Markdown
> and choose which nodes to recommission.
> 
> Recently dead nodes, that hold unexpired leases, may cause decommissioning to hang until `COCKROACH_TIME_UNTIL_STORE_DEAD` elapses.
> ```

It's not really about leases but the fact that we don't move data off until `COCKROACH_TIME_UNTIL_STORE_DEAD` has passed.

---

*[docs/RFCS/decommission.md, line 153 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km3I4HBfDLft6zs_ZS9:-Km3I4HBfDLft6zs_ZSA:bid4hf7) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee9d6a23e11b7d157725bd45bf860116b393869/docs/RFCS/decommission.md#L153)):*
> ```Markdown
> Recently dead nodes, that hold unexpired leases, may cause decommissioning to hang until `COCKROACH_TIME_UNTIL_STORE_DEAD` elapses.
> 
> # Alternatives
> ```

Move this to the section where you discuss temporary outages. Make that section part of the motivation (as in: Note that the case in which you temporarily want to remove a node from the cluster for maintenance is sufficiently covered by increasing `TIME_UNTIL_STORE_DEAD`, which should become a cluster setting in the foreseeable future.), etc. That motivates what you want to solve here and what not.

---

*[docs/RFCS/decommission.md, line 163 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km3IMtLzxZ2oXAAlhiS:-Km3IMtMJlr-Yuq-tVe2:b-z7f17y) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee9d6a23e11b7d157725bd45bf860116b393869/docs/RFCS/decommission.md#L163)):*
> ```Markdown
> 
> # Unresolved questions
> - No cosmetic change to the UI is required: we want to know which nodes are part
> ```

Move this to a section in the actual design: The UI hides nodes which are marked as dead and are also in decommissioning state (as per the liveness table (or rather its gossiped information)). That should have the desired effect and is straightforward while keeping stats, etc.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-:-Km3IkntXiXSEEOEo5KJ:b8iezok)*
<!-- Sent from Reviewable.io -->

neeral committed:
Add RFC for decommissioning a node

neeral commented:
Review status: 0 of 1 files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/decommission.md, line 54 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Kly3AXKlLOYbTK-8orG:-Km7C1-iYypk_s7X-hU3:b-qppx9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b48641ccecd1bd9f29369c1621c4bf1f40e6a6/docs/RFCS/decommission.md#L54)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think the command should block, but be retryable (so if the command fails, just run `cockroach node rm` again). The command exits when it is safe to wipe/power down the machine. 
</blockquote></details>

There are two variants of the command. One is non-blocking and one is blocking (this also performs node shutdown).

---

*[docs/RFCS/decommission.md, line 91 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km3FmAzkMhzn_3_F5_f:-Km7BSUqdMVURIZJrD_l:bj4ob47) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee9d6a23e11b7d157725bd45bf860116b393869/docs/RFCS/decommission.md#L91)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I think you *could* do better here. If the node is live, you could directly tell it to undrain. But I think it's OK to punt on this for now. Perhaps make a note.
</blockquote></details>

I have added this to the Drawbacks section and made a note of this as future work.

---

*[docs/RFCS/decommission.md, line 147 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km35RNZ809C-MKFDd4A:-Km7pPnJreS2jgkBA7Le:b-6d6tkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee9d6a23e11b7d157725bd45bf860116b393869/docs/RFCS/decommission.md#L147)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Yeah, that's not really a drawback of the algorithm here unless you're trying to make all invocations of `decommission` either succeed in bounded time or fail. The cluster will simply refuse to decommission itself to death, and decommissioning will hang. Nice to detect, but with an effort that I think is disproportionate related to the win it gives in the short term.
</blockquote></details>

I've updated the description according to Ben's corrections.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
LGTM. I'll check in with you if you have any further changes queued and then merge this.

---

Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/decommission.md, line 24 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km8B_rK4ZZIFoyt484P:-Km8B_rK4ZZIFoyt484Q:b-5m1ime) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bac1d6e3ec14e32d2e4164f4617fbfa41c66492/docs/RFCS/decommission.md#L24)):*
> ```Markdown
> re-replicating previously deleted entries. CockroachDB does not suffer from this
> specific issue.
> A typical operation in the field is to replacing a set of nodes. This involves
> ```

s/replacing/replace/

---

*[docs/RFCS/decommission.md, line 156 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-Km8C4JRt00iDXlwGKtW:-Km8C4JRt00iDXlwGKtX:b1f6rfy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bac1d6e3ec14e32d2e4164f4617fbfa41c66492/docs/RFCS/decommission.md#L156)):*
> ```Markdown
>   flag.
>   - *Nice to have*: satisfying safety constraints as a pre-requisite. As an
>     example: `number_of_nodes < max(zone.number_of_replicas_desired`. It would
> ```

![✂-1](https://files.reviewableusercontent.io/images/reviewable/5076964/8-zxCK82IJ44LwHC1ZAirtaM.png)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-:-Km8CZSTcyZHIpF2i5MP:bqhb2dn)*
<!-- Sent from Reviewable.io -->

neeral commented:
Thanks Ben, Peter and Toby for reviewing my first RFC.

---

Review status: 0 of 1 files reviewed at latest revision, 19 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-:-Km8IsJOxQRTITx5R0AT:b-tbz38u)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great first contribution, @neeral!

---

Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16352#-:-KmD4J7YOIxlGFRCtRjr:byckep4)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Closing since @neeral accidentally removed their fork. We'll reopen and merge this shortly.

EamonZhang commented:
mark
